### PR TITLE
Fix Partial WebSocket Reads

### PR DESCRIPTION
**Pull request by zeroSteiner**
BinData does not [handle](https://github.com/dmendel/bindata/blob/de567f8cbdc6db612b51fe27eddf4c358548ce08/lib/bindata/io.rb#L313) instances where an object's `#read` method returns less data than was anticipated which is the case of Metasploit's sockets. This bug manifests itself when a large chunk of data is received by the websocket and ultimately causes it to be closed with an `IOError`.

This change updates the `get_wsframe` method to read the header and payload length from the socket, then read the payload data, accounting for any partial reads. Data is read until the expected length of the payload is met. This does mean that the header fields of the WebSocket `Frame` object have to be broken out to be accessible on their own which in turn lead to a bit of refactoring. I also ran `rubocop -a` on the library file which changed a whole bunch of things.

## Verification

List the steps needed to make sure this thing works

- [ ] Create the kubernetes environment
```
cd metasploit-framework/test/kubernetes
make thinkphp
make forward-thinkphp
```

- [ ] Exploit:
```
use unix/webapp/thinkphp_rce
run http://192.168.123.1:9001 lhost=x.y.x.z
```

- [ ] Use the kubernetes exec module:
```
use exploit/multi/kubernetes/exec
run session=-1 lhost=x.y.x.z podimage=vulhub/thinkphp:5.0.23
```

- [ ] In the new session run the following command, see 'hello world' times 1000
    * Prior to these changes, the session would abruptly close due to the IOError
```
php -r 'echo str_repeat("hello world", 1000);'
mount
```

### CODE REVIEW

File: lib/rex/proto/http/web_socket.rb [hunk 1]
```diff
@@ -1,428 +1,445 @@
...
+          # Read a WebSocket::Frame from the peer.
+          #
+          # @return [WebSocket::Frame] the frame that was received from the peer.
+          def get_wsframe(_opts = {})
+            frame = Frame.new
+            frame.header.read(self)
+            payload_data = ''
+            while payload_data.length < frame.payload_len
+              chunk = read(frame.payload_len - payload_data.length)
+              return nil if chunk.nil? # no partial reads!
```
Comment (dwelch-r7) [inline lib/rex/proto/http/web_socket.rb:200]: Do you know what can cause the partial reads?
Would it be worth logging out that one happened?

File: lib/rex/proto/http/web_socket.rb [hunk 2]
```diff
@@ -1,428 +1,449 @@
...
-        websocket.wsloop do |data, data_type|
-          next unless @read_type.nil? || data_type == @read_type
-
-          data = on_data_read(data, data_type)
-          next if data.nil?
-
-          rsock.syswrite(data)
+  module Proto
+    module Http
+      module WebSocket
```
Comment (adfoster-r7) [inline lib/rex/proto/http/web_socket.rb:8]:
```suggestion
module Rex::Proto::Http::WebSocket
```

Which might clean up the indentation diff

Comment (smcintyre-r7) [reply lib/rex/proto/http/web_socket.rb]: Well I did that, but to keep rubocop happy still, the indentation still changed.

**Comment by adfoster-r7** [comment]
## Release Notes

Fixes an edgecase in the Kubernetes exec module which led to sessions dying when performing partial websocket reads

**Pull request merged**